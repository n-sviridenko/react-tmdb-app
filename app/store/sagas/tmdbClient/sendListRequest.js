import { call } from 'redux-saga/effects';
import { Map } from 'immutable';

import * as schemas from 'store/schemas';
import sendRequest from './sendRequest';

function* sendListRequest(methodName, itemSchema, ...args) {
  const res = yield call(sendRequest, methodName, schemas.makeList(itemSchema), ...args);

  return Map({
    count: res.get('total_results'),
    pageCount: res.get('total_pages'),
    page: res.get('page'),
    items: res.get('results'),
  });
}

export default sendListRequest;
