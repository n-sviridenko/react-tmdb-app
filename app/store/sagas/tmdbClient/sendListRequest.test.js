/* eslint-disable redux-saga/yield-effects */
import { fromJS } from 'immutable';

import sendListRequest from './sendListRequest';

describe('tmdbClient sendListRequest saga', () => {
  it('should dispatch call sendRequest and normalize the response', () => {
    const task = sendListRequest('foo.bar', {}, 'a', 'b');
    const res = fromJS({ total_results: 10, total_pages: 2, page: 1, results: [] });
    const list = fromJS({ count: 10, pageCount: 2, page: 1, items: [] });
    const callDescriptor = task.next().value;
    expect(callDescriptor).toMatchSnapshot();
    const returnDescriptor = task.next(res).value;
    expect(returnDescriptor).toEqual(list);
  });
});
