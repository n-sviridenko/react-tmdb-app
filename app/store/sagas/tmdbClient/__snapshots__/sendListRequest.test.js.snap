exports[`tmdbClient sendListRequest saga should dispatch call sendRequest and normalize the response 1`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "foo.bar",
      Object {
        "results": ArraySchema {
          "schema": Object {},
        },
      },
      "a",
      "b",
    ],
    "context": null,
    "fn": [Function],
  },
}
`;
